glean-core/src/upload/directory.rs

```diff
@@ -175,10 +180,35 @@ impl PingDirectoryManager {
         });
 
         // Return the vector leaving only the `PingRequest`s in it
-        Ok(pending_pings
+        pending_pings
             .into_iter()
             .map(|(_, request)| request)
-            .collect())
+            .collect()
+    }
+
+    /// Get all the ping entries in all ping directories.
+    fn get_ping_entries(&self) -> Vec<fs::DirEntry> {
+        let mut result = Vec::new();
+        for dir in &self.get_dirs() {
+            if let Ok(entries) = dir.read_dir() {
+                result.extend(entries.filter_map(|entry| entry.ok()))
+            };
+        }
+        result
+    }
+
+    /// Get the path for a ping file based on its uuid.
+    ///
+    /// Will look for files in each ping directory until something is found.
+    /// If nothing is found, returns `None`.
+    fn get_file_path(&self, uuid: &str) -> Option<PathBuf> {
```

Comment:
I did this so that we didn't need to always know which kind of ping it is we are searching for. 

Since most times we will be looking for ordinary pings, it doesn't really change much, we never even go into the deletion_request directory (I was careful about ordering the dirs slice to have the pending_pings folder path first). 

If we are indeed looking for a deletion-request ping file, then it will scan the pending_pings directory before going into the deletion_request directory, which is not ideal but shouldn't be a big problem, since the pending_pings directory will be (or should be) empty by then.

Reply:
> Since most times we will be looking for ordinary pings, it doesn't really change much, we never even go into the deletion_request directory (I was careful about ordering the dirs slice to have the pending_pings directory path first).

At the very least, we should document this requirement where you are declaring the dirs slice.

> If we are indeed looking for a deletion-request ping file, then it will scan the pending_pings directory before going into the deletion_request directory, which is not ideal but shouldn't be a big problem, since the pending_pings directory will be (or should be) empty by then.

Does this really matter? Why is order important in this case?

I think we should probably do a better job at _not_ sending data if upload is off. Removing non-'deletion' pings can still happen while the product asks Glean for things to upload, meaning we could be sending pings even though upload was off, because of racing. Thoughts?

Reply:
> Why is order important in this case?

It is important based on the assumption that most pings we send are not deletion-request pings. Order is important to make sure we scan pending_pings directory before falling back to deletion_request directory, thus ensuring we don't perform the usually pointless step of reading the deletion_request directory.

Reply:
> 
> 
> > Why is order important in this case?
> 
> It is important based on the assumption that most pings we send are not deletion-request pings. Order is important to make sure we scan pending_pings directory before falling back to deletion_request directory, thus ensuring we don't perform the usually pointless step of reading the deletion_request directory.

So it is important because of performance? Not because we break functionality?

Reply:
> I think we should probably do a better job at not sending data if upload is off. Removing non-'deletion' pings can still happen while the product asks Glean for things to upload, meaning we could be sending pings even though upload was off, because of racing. Thoughts?

This comment has made me go back and look again at the steps in `set_upload_enabled`. It goes like this:
1. Submit the deletion-request ping;
2. Change upload enabled flag;
3. Clear pending_pings_directory;
4. Clear stored metrics;
5. etc;

I guess if it is possible to have a race condition there and get a ping to send before finishing clearing all metrics on the rust side, we have that problem already.

In android at least we lock the the PingUploader, so that we don't iterate over pings while deleting them.

I guess I could in some way lock the pings queue from the moment we start clearing all the stores and queues until we finish it, so that we don't need to rely on the platform side doing that, but that should happen when I add the upload manager functionality to the Glean object, which will happen in my next PR.

Reply:
> So it is important because of performance? Not because we break functionality?

Yes, we skip some steps this way. So you could say, performance :)

Reply:
Ok, setting aside the race from the previous comment, if performance is the only remaining issue, I think it's a minor one: we expect this to be off the main thread. While this still means I/O, it doesn't necessarily mean frequent.

Reply:
> I guess I could in some way lock the pings queue from the moment we start clearing all the stores and queues until we finish it, so that we don't need to rely on the platform side doing that, but that should happen when I add the upload manager functionality to the Glean object, which will happen in my next PR.

Let's add this as a TODO here (in a comment) and mention when it will happen (bug number).

Reply:
Hm,  I am not sure that it makes sense to add a comment like this here. This function is considerably distant from where that code will be.

Since implementing this is what I will do immediatelly after closing this PR, do you agree to just add a comment to that bug, so that we know that is one important thing to do on it? (I did that already, btw. https://bugzilla.mozilla.org/show_bug.cgi?id=1614284#c1)